code/perception/src/segmentation_node.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+
+import pathlib
+import numpy as np
+import ros_compatibility as roscomp
+import torch
+import cv2
+import albumentations as A
+from panopticapi.utils import id2rgb
+from rospy.numpy_msg import numpy_msg
+from ros_compatibility.node import CompatibleNode
+from sensor_msgs.msg import Image
+
+from panoptic_segmentation.efficientps import EfficientPS as EfficientPS
+from panoptic_segmentation.train_net import add_custom_param
+from panoptic_segmentation.efficientps.panoptic_segmentation_module import \
+    panoptic_segmentation_module
+
+from detectron2.config import get_cfg
+import torchvision.transforms.functional as F
+from detectron2.structures import Instances, BitMasks, Boxes
+
+CFG_FILE_PATH = pathlib.Path(
+    __file__).parent / "panoptic_segmentation" / "config.yaml"
+MODEL_PATH = pathlib.Path(
+    __file__).parent.parent / \
+             "models/panoptic_segmentation/efficientps.ckpt"
+
+
+class SegmentationNode(CompatibleNode):
+    """
+    This node runs the panoptic segmentation model on
+    the camera images and publishes the segmented results.
+    """
+
+    def __init__(self, name, **kwargs):
+        super().__init__(name, **kwargs)
+
+        self.publisher = None
+        self.loginfo("Initializing panoptic segmentation node...")
+
+        self.role_name = self.get_param("role_name", "hero")
+        self.side = self.get_param("side", "Center")
+
+        self.model, self.transform, self.model_cfg = self.load_model()
+        # warm up
+        self.predict(np.zeros((720, 1280, 3)))
+
+        self.setup_camera_subscriptions()
+        self.setup_camera_publishers()
+
+    def setup_camera_subscriptions(self):
+        self.new_subscription(
+            msg_type=numpy_msg(Image),
+            callback=self.handle_camera_image,
+            topic=f"/carla/{self.role_name}/{self.side}/image",
+            qos_profile=1
+        )
+
+    def setup_camera_publishers(self):
+        self.publisher = self.new_publisher(
+            msg_type=numpy_msg(Image),
+            topic=f"/paf/{self.role_name}/{self.side}/segmented_image",
+            qos_profile=1
+        )
```

Comment:
As far as I understood, the segmentation node only publishes the segmentated images, right? Where is the detection of traffic lights, traffic signs, cars driving in front of us, etc. handled?

Reply:
These features aren't implemented yet and not part of this PR. I think, I will create another issue for traffic light snipping and classification.